compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/optimizer/BIRLockOptimizer.java

```diff
@@ -0,0 +1,256 @@
+    private Map<BIRTerminator.Lock, Integer> lockToSetMap = new HashMap<>();
+    private Map<Integer, List<BIRTerminator.Lock>> setToLockMap = new HashMap<>();
+    private int setId = -1;
+
+    public void optimizeNode(BIRNode node) {
+        // Collect lock nodes
+        node.accept(this);
+
+        // Identify disjoint locks
+        optimizeLocks();
+    }
+
+    private void optimizeLocks() {
+        analyzeLocks();
+        propagateLocks();
+    }
+
+    private void propagateLocks() {
+        for (Map.Entry<Integer, List<BIRTerminator.Lock>> entry : setToLockMap.entrySet()) {
+            Integer lockId = entry.getKey();
+            for (BIRTerminator.Lock lock : entry.getValue()) {
+                lock.lockId = lockId;
+            }
+        }
+    }
+
+    private void analyzeLocks() {
+        for (int lockListIndex = 0; lockListIndex < lockList.size(); lockListIndex++) {
+            if (!lockToSetMap.containsKey(lockList.get(lockListIndex))) {
+                analyzeUnvisitedLock(lockListIndex);
+            } else {
+                analyzeVisitedLock(lockListIndex);
+            }
+        }
+    }
+
+    private void analyzeVisitedLock(int lockListIndex) {
+        BIRTerminator.Lock currentLock = lockList.get(lockListIndex);
+        int previousSetId = setId;
+        Integer currentSetIdLocal = lockToSetMap.get(currentLock);
+        List<BIRTerminator.Lock> currentSet = setToLockMap.get(currentSetIdLocal);
+
+        // Use the setId of the current one for the other shared locks
+        setId = currentSetIdLocal;
+        populateLockSet(currentSet, currentLock, lockListIndex);
+        setId = previousSetId;
+    }
+
+    private void analyzeUnvisitedLock(int lockListIndex) {
+        BIRTerminator.Lock currentLock = lockList.get(lockListIndex);
+        List<BIRTerminator.Lock> currentSet = new LinkedList<>();
+
+        // Add to the maps for the current unvisited lock.
+        lockToSetMap.put(currentLock, ++setId);
+
+        // Compare with the rest of the locks.
+        currentSet.add(currentLock);
+
+        populateLockSet(currentSet, currentLock, lockListIndex);
+
+        setToLockMap.put(setId, currentSet);
+    }
+
+    private void populateLockSet(List<BIRTerminator.Lock> currentSet, BIRTerminator.Lock currentLock,
+            int lockListIndex) {
+        for (int i = (lockListIndex + 1); i < lockList.size(); i++) {
+            BIRTerminator.Lock comparedLock = lockList.get(i);
+            Set<BIRNode.BIRGlobalVariableDcl> globalVarSetOfComparedLock = comparedLock.lockVariables;
+
+            if (isSharedLock(currentLock, globalVarSetOfComparedLock)
+                && isNotInSameSet(currentLock, comparedLock)) {
+                populateLockSet(currentSet, comparedLock);
+            }
+        }
+    }
+
+    private boolean isNotInSameSet(BIRTerminator.Lock currentLock, BIRTerminator.Lock comparedLock) {
+        Integer currentLockId = lockToSetMap.get(currentLock);
```

Comment:
should this be `currentLockSetId` instead of `currentLockId ` ?

Reply:
I think its better. I'll change it